Mafia update: Return to body when the game ends

JohnFulpWillard opened the pull request:
## About The Pull Request

Mafia players are now sent to their last body when the Mafia game ends, and in the meantime they have text saying why they are dead.
They can still be revived during this period, and when the game is up they'll be forced back into their body if alive, if they are dead then they'll just be able to re-enter their corpse as normal.

Also since I was poking around in Mafia stuff:
- I removed mafia observing because it was unused (this previously was used to allow ghosts to see Changeling chat).
- I fixed it being a Draw when there was one Town or Changeling left.
- I fixed the role list showing some roles multiple times
- I fixed the Chaplain not being able to use their night ability (and therefore being completely useless)
- I added prevention to prevent Admins from causing runtimes or straight up crashing the server, with a very real chance it can happen purely by accident, through the Admin UI.

I'm hoping to change how this actually does the job because I find it to be very bad coding practices, but my problem is that everyone who signs up for Mafia is a ghost, and they are added into the game through their CLIENTS, so we don't have access on the mafia controller or the role, to the player's previous body or mind, without this shit.

Also adds a new mafia board icon
![board](https://user-images.githubusercontent.com/53777086/236792499-98ca2949-6e20-4d4f-9bee-046ff7345589.png)

in-game (the video is wrong and was an error due to importing from png)
![image](https://user-images.githubusercontent.com/53777086/236792967-c6ead5b3-25ed-405c-bee6-0d7689f39f78.png)

Made by tatax and I find it fits more the theme of Mafia than the current one.

New UI
![image](https://user-images.githubusercontent.com/53777086/236939302-ffead270-3b5b-48ae-8775-b38531a2055c.png)

### Video demonstration

https://cdn.discordapp.com/attachments/1057125273369329754/1105069159408210000/2023-05-08_05-38-53.mp4

## Why It's Good For The Game

Players can freely sign up for Mafia without having to worry that they'll be DNR'ed for the rest of the game, and players will know why they are temporarily AFK.

Closes https://github.com/tgstation/tgstation/issues/64282

## Changelog

:cl: JohnFulpWillard, tatax
qol: [Mafia] Players are now sent back to their in-game character's bodies when they finish a game.
image: [Mafia] Mafia now has a new sprite for the main action button in-game.
fix: [Mafia] Chaplains can now use their ability.
fix: [Mafia] The game no longer ends in a Draw if Town or Changelings win.
fix: [Mafia] The role list now properly shows roles once.
/:cl:

JohnFulpWillard committed:
Mafia players are put back in their bodies

JohnFulpWillard committed:
Don't prevent re-entering your body

DATA-xPUNGED commented:
Will this happen for other ghost minigames?

JohnFulpWillard commented:
I dont currently have plans to mess with other mini games and they don't currently share any code (they're all they're own datum) to generalize it into.

BlueMemesauce commented:
https://user-images.githubusercontent.com/47338680/236803400-fa09fe58-99b3-4c15-859a-07121929ce41.mp4

Can you upload your videos directly to github? It may be easier for you to just put the link, but the rest of us have to copy paste it over to discord to see it embed or we have to download the file ourselves. If the video needs to be compressed, the https://8mb.video/ site is pretty great for this, but the HandBrake https://github.com/HandBrake/HandBrake open-sourced software can also be used.

 It's also more likely that a discord link will die compared to a github link because it will die if anyone ever deletes the message it's gone. Although if you posted it in some random DM then it's probably safe as the only person who can ever delete it is you.

JohnFulpWillard committed:
Grab the ghost if they aren't in their body

JohnFulpWillard committed:
UI improvements

JohnFulpWillard committed:
Fixes chaplain

JohnFulpWillard committed:
Fixes other Mafia issues

Mafia no longer runtimes trying to send the next phase before a game starts
Admins no longer crash the server when Mafia starts and they were doing a custom setup
Role list no longer shows multiple roles several times

san7890 commented on code/modules/mafia/roles/roles.dm:
```diff
@@ -46,11 +51,43 @@
+ * If the player is alive again, we'll shove them back in.
+ */
+/datum/mafia_role/proc/send_back_to_body()
+	if(old_mind && !QDELETED(old_mind) && old_body && !QDELETED(old_body))
```
QDELETED does a nullcheck, so this is just pretty redundant, should be:

```suggestion
	if(!QDELETED(old_mind) && !QDELETED(old_body))
```

also this could probably be an early return instead to save that indentation

MrMelbert reviewed (changes requested):
TBQH I would greater prefer this system be generalized rather than isolated to only Mafia
Not only could we use it for the other mini-games but it would just be a very useful tool when handling "temporarily" stealing a player's mind / ghost to do something else

Quite simply it can be made a component applied to the mob or something

JohnFulpWillard committed:
Component & other reviews

JohnFulpWillard committed:
Uses weakrefs instead like a boss

JohnFulpWillard commented on code/modules/mafia/roles/roles.dm:
```diff
@@ -46,11 +51,43 @@
 		role_unique_actions -= abilities
 
 /datum/mafia_role/Destroy(force, ...)
+	send_back_to_body()
```
I'm just not too entirely sure at what point we should have it in. I was thinking of putting it in mafia's ``end_game()``, but now since it's a component I'm not sure it'll work out that well.

JohnFulpWillard committed:
Update temporary_body.dm

JohnFulpWillard committed:
wtf lol

MrMelbert commented on code/modules/mafia/roles/roles.dm:
```diff
@@ -48,9 +48,31 @@
+/datum/mafia_role/proc/put_player_in_body(client/player)
+	if(player.mob.mind && player.mob.mind.current)
+		ADD_TRAIT(player.mob.mind.current, TRAIT_PLAYING_MAFIA, MAFIA_TRAIT)
+		body.AddComponent(/datum/component/temporary_body, \
+			/* old_mind = */ player.mob.mind, \
+			/* old_body = */ player.mob.mind.current, \
+			/* body_return_callback = */ CALLBACK(src, PROC_REF(remove_mafia_trait)), \
+		)
```
you can use named args for add component, why beat around the bush

MrMelbert commented on code/modules/mafia/controller.dm:
```diff
@@ -337,7 +337,7 @@ GLOBAL_LIST_INIT(mafia_role_by_alignment, setup_mafia_role_by_alignment())
 	var/victory_message
 
-	if((living_mafia.len + living_town.len) <= 1)
+	if(!(living_mafia.len + living_town.len))
```
the older format was a lot clearer

MrMelbert commented on code/modules/mafia/roles/roles.dm:
```diff
@@ -48,9 +48,31 @@
+ * Send by a callback from the temporary body component
+ * Used to clear the playing mafia trait from their old body.
+ */
+/datum/mafia_role/proc/remove_mafia_trait(datum/mind/old_mind, mob/living/old_body)
+	REMOVE_TRAIT(old_body, TRAIT_PLAYING_MAFIA, MAFIA_TRAIT)
```
You can just use `TRAIT_CALLBACK_REMOVE` instead of defining your own proc

MrMelbert commented on code/modules/mafia/roles/roles.dm:
```diff
@@ -48,9 +48,31 @@
+/datum/mafia_role/proc/put_player_in_body(client/player)
+	if(player.mob.mind && player.mob.mind.current)
+		ADD_TRAIT(player.mob.mind.current, TRAIT_PLAYING_MAFIA, MAFIA_TRAIT)
+		body.AddComponent(/datum/component/temporary_body, \
```
```suggestion
		body.AddComponent( \
			/datum/component/temporary_body, \
```

JohnFulpWillard committed:
Merge branch 'master' into mafia-reenter-body

JohnFulpWillard committed:
Merge branch 'mafia-reenter-body' of https://github.com/JohnFulpWillard/tgstation into mafia-reenter-body

JohnFulpWillard committed:
Generalizes the trait and other melbert review comments

JohnFulpWillard commented on code/modules/mob/living/carbon/human/examine.dm:
```diff
@@ -121,6 +121,8 @@
 			if(HAS_TRAIT(src, TRAIT_SUICIDED))
 				. += span_warning("[t_He] appear[p_s()] to have committed suicide... there is no hope of recovery.")
+			if(HAS_TRAIT(src, TRAIT_PLAYING_MAFIA))
+				. += span_warning("[t_He] appear[p_s()] to still be lingering somewhere... [p_they()] will surely be back soon.")
```
I decided to just not have any text so people don't confuse it for the suicide text.
I remember last time we changed them and it was reverted because people couldn't tell the difference.

JohnFulpWillard commented:
@MrMelbert Is this PR good to go?
